### DISCUSSION TITLE
Cart / Checkout Implementation Details

### DISCUSSION THREAD

**Discussion by apbendi** [dcgtc/dgrants#90] [Architecture & Design] [answered]
* When a user clicks "Add to Cart", what does the app do?
    *  Add data to local storage and/or IndexedDB?
    * What data do we store— the grantId, supplied token address, and amount?
    * What supply token do we default to? What amount?
    * Do we have a side cart of any kind like cGrants?
* Do we have a sidecart?
* Do we want some/all of the token selection mechanics that exist on the cGrants cart?
* What other considerations exist for the cart implementation?

**Comment by thelostone-mc** [comment]
> What data do we store— the grantId, supplied token address, and amount?

The grantRoundManager? Well that would be handled by dApp itself

---

> What supply token do we default to? What amount?

5 DAI ? I say this cause this was the default DAI valus on dGrants and DAI is commonly used.
The smarter option would be let the dApp keep it the same as the donationToken (but in our case GTC isn't actively traded so might add friction for the default user)

---

> Do we have a sidecart?

Looking at @scco design, I'd say no. Might be worth doing a design walkthrough with the entire team 

---

> Do we want some/all of the token selection mechanics that exist on the cGrants cart?

It would be just the token registry availabe on uniswap right ?

---

> What other considerations exist for the cart implementation?

What happens once we sign the transaction ? Would the user wait until txn is completed / what do do we do if txn fails ?

**Reply by mds1** [reply]
> What happens once we sign the transaction ? Would the user wait until txn is completed / what do do we do if txn fails ?

I think we wait until the transaction is completed, and show the "thanks for checking out modal" when done and redirect to another page. If it fails, keep them on the cart page, show the error, and let them know they'll need to check out again

**Comment by mds1** [comment]
> Add data to local storage and/or IndexedDB?

Local storage should be sufficient for now and we can migrate to IndexedDB later if we need to. Perhaps we pull the logic of reading/writing to local storage into some utility methods so we can easily replace it with IndexedDB if necessary, cc @gdixon 

> What data do we store— the grantId, supplied token address, and amount?

I think this should be all we need to store, and the cart will have to fetch the rest of the grant data based on the ID whenever it's loaded. For now looking that up explicitly each time seems fine, but we'll probably want to add some caching to this app at some point

> What supply token do we default to? What amount?

Agree with @thelostone-mc that 5 DAI seems reasonable for now

> Do we have a sidecart?

Personally I never saw much value in the side-cart and would be in favor of not having one as it keeps things simpler and probably isn't a priority for the MVP

> Do we want some/all of the token selection mechanics that exist on the cGrants cart?

Do you mean the dropdown in the cart where you select the token? If so, I think we pick the top 5 most popular cgrants tokens and only support those to start

> What other considerations exist for the cart implementation?

1. cgrants has an optional tip to Gitcoin which defaults to 5%. Do we want an optional tip? And if so who should it go to? The DAO?
2. When checking out, cgrants has a "thanks, here is a link to share your cart" modal. I don't know if we need the share link for the MVP, but some "thank you" modal + redirect after a successful checkout is probably nice to have

**Reply by phutchins** [reply]
> cgrants has an optional tip to Gitcoin which defaults to 5%. Do we want an optional tip? And if so who should it go to? The DAO?

It would be great to have the option to tip the Entity running the explorer. We could have it default to funding the decentralization of Gitcoin. I.E. the dGTC multisig (0x931896A8A9313F622a2AFCA76d1471B97955e551).

> When checking out, cgrants has a "thanks, here is a link to share your cart" modal. I don't know if we need the share link for the MVP, but some "thank you" modal + redirect after a successful checkout is probably nice to have

It would be great if not too difficult to have the ability to share "I just donated to X grant on decentralized gitcoin grants" to help promote the platform. I would say that it is probably bottom of the list as far as priorities for the initial dGrants round.

**Reply by apbendi** [reply]
> It would be great to have the option to tip the Entity running the explorer. We could have it default to funding the decentralization of Gitcoin. I.E. the dGTC multisig (0x931896A8A9313F622a2AFCA76d1471B97955e551).

This can/should probably be a separate discussion/issue/feature from the actual cart checkout flow.

> It would be great if not too difficult to have the ability to share "I just donated to X grant on decentralized gitcoin grants" to help promote the platform. I would say that it is probably bottom of the list as far as priorities for the initial dGrants round.

Similarly, let's track this as a separate feature request and keep it out of the critical path for getting the checkout cart working.

**Comment by apbendi** [comment, marked as answer]
These questions were answered in #94 and implemented in #111
